nomad/job_endpoint.go

```diff
@@ -783,44 +824,25 @@ func (j *Job) Deregister(args *structs.JobDeregisterRequest, reply *structs.JobD
 		result = multierror.Append(result, err)
 	}
 
-	// If the job is periodic or parameterized, we don't create an eval.
-	if job != nil && (job.IsPeriodic() || job.IsParameterized()) {
-		return nil
-	}
-
-	// Create a new evaluation
-	// XXX: The job priority / type is strange for this, since it's not a high
-	// priority even if the job was.
-	now := time.Now().UTC().UnixNano()
-	eval := &structs.Evaluation{
-		ID:             uuid.Generate(),
-		Namespace:      args.RequestNamespace(),
-		Priority:       structs.JobDefaultPriority,
-		Type:           structs.JobTypeService,
-		TriggeredBy:    structs.EvalTriggerJobDeregister,
-		JobID:          args.JobID,
-		JobModifyIndex: index,
-		Status:         structs.EvalStatusPending,
-		CreateTime:     now,
-		ModifyTime:     now,
-	}
-	update := &structs.EvalUpdateRequest{
-		Evals:        []*structs.Evaluation{eval},
-		WriteRequest: structs.WriteRequest{Region: args.Region},
-	}
+	// COMPAT(1.1.0) - 0.12.1 introduced atomic job deregistration eval
+	if args.Eval != nil &&
+		!ServersMeetMinimumVersion(j.srv.Members(), minJobRegisterAtomicEvalVersion, false) {
+		// Create a new evaluation
+		args.Eval.JobModifyIndex = index
+		update := &structs.EvalUpdateRequest{
+			Evals:        []*structs.Evaluation{args.Eval},
+			WriteRequest: structs.WriteRequest{Region: args.Region},
+		}
 
-	// Commit this evaluation via Raft
-	_, evalIndex, err := j.srv.raftApply(structs.EvalUpdateRequestType, update)
-	if err != nil {
-		result = multierror.Append(result, err)
-		j.logger.Error("eval create failed", "error", err, "method", "deregister")
-		return result.ErrorOrNil()
+		// Commit this evaluation via Raft
+		_, _, err := j.srv.raftApply(structs.EvalUpdateRequestType, update)
```

Comment:
We're not using the `evalIndex` returned here to set the `reply.EvalCreateIndex` but is the index returned from registration really the right index if we're on this branch? Compare to what we've done here in `Job.Register`.

Reply:
I think so.  This path is relevant when the leader is 0.12.1 but some followers are pre-0.12.1.  In this case, the servers will disagree on eval.CreateIndex and I'm opting to return the value as seen by the leader and will be returned in subsequent calls (notwithstanding more leader elections).  This disagreement doesn't seem material or significant and only affects the evals created during the mixed cluster stage, so I'm willing to keep it as-is without worrying about it much.  I called this out in the PR description but didn't highlight the implication to this.

The job registration is a bit tricky, and I think this behavior matches the one there too.  In job registration, we only use update the evalIndex if it's not a duplicate compatibility eval, due to the `evalSubmitted` check: https://github.com/hashicorp/nomad/blob/97c69ee9a7312aaa315fec5850fe0a58e13a238a/nomad/job_endpoint.go#L392-L395.

Reply:
> This disagreement doesn't seem material or significant and only affects the evals created during the mixed cluster stage, so I'm willing to keep it as-is without worrying about it much.

Fair enough. We don't use that index for any kind of check-and-set operation as far as I know.

> The job registration is a bit tricky, and I think this behavior matches the one there too. In job registration, we only use update the evalIndex if it's not a duplicate compatibility eval, due to the `evalSubmitted` check:

Ok, and as you mentioned in https://github.com/hashicorp/nomad/pull/8435/files#r454546652 we need to keep that check even after we remove the backwards compatibility. This makes sense to me. :+1: